File: lever/src/main/java/com/pin/lever/ResourceRequestHandler.java
package com.pin.lever;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;

import java.io.IOException;

import static android.content.ContentResolver.SCHEME_ANDROID_RESOURCE;
import static com.pin.lever.Lever.LoadedFrom.DISK;


class ResourceRequestHandler extends RequestHandler {
    private final Context context;

    ResourceRequestHandler(Context context) {
        this.context = context;
    }

    @Override public boolean canHandleRequest(Request data) {
        if (data.resourceId != 0) {
            return true;
        }

        return SCHEME_ANDROID_RESOURCE.equals(data.uri.getScheme());
    }

    @Override public Result load(Request request, int networkPolicy) throws IOException {
        Resources res = Utils.getResources(context, request);
        int id = Utils.getResourceId(res, request);
        return new Result(decodeResource(res, id, request), DISK);
    }

    private static Bitmap decodeResource(Resources resources, int id, Request data) {
        final BitmapFactory.Options options = createBitmapOptions(data);
        if (requiresInSampleSize(options)) {
            BitmapFactory.decodeResource(resources, id, options);
            calculateInSampleSize(data.targetWidth, data.targetHeight, options, data);
        }
        return BitmapFactory.decodeResource(resources, id, options);
    }
}

File: lever/src/main/java/com/pin/lever/ResourceRequestHandler.kt
package com.pin.lever

import android.content.ContentResolver
import android.content.Context
import android.content.res.Resources
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.pin.lever.Lever.LoadedFrom
import com.pin.lever.Utils.getResourceId
import com.pin.lever.Utils.getResources
import java.io.IOException

internal class ResourceRequestHandler(private val context: Context) :
    RequestHandler() {
    override fun canHandleRequest(data: Request?): Boolean {
        return if (data!!.resourceId != 0) {
            true
        } else ContentResolver.SCHEME_ANDROID_RESOURCE == data.uri!!.scheme
    }

    @Throws(IOException::class)
    override fun load(
        request: Request?,
        networkPolicy: Int
    ): Result? {
        val res = getResources(context, request!!)
        val id = getResourceId(res, request)
        return Result(
            decodeResource(
                res,
                id,
                request
            ), LoadedFrom.DISK
        )
    }

    companion object {
        private fun decodeResource(
            resources: Resources,
            id: Int,
            data: Request
        ): Bitmap {
            val options = createBitmapOptions(data)
            if (requiresInSampleSize(options)) {
                BitmapFactory.decodeResource(resources, id, options)
                calculateInSampleSize(
                    data.targetWidth,
                    data.targetHeight,
                    options!!,
                    data
                )
            }
            return BitmapFactory.decodeResource(resources, id, options)
        }
    }

}